src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1096 @@
+ * tasks recorded in pg_dist_background_tasks and ensures execution based
+ * on a statemachine. When a task needs to be executed it starts a
+ * Background Task Executor that executes the sql statement defined in the
+ * task. The output of the Executor is shared with the Monitor via a
+ * shared memory queue.
+ *
+ * To make sure there is only ever exactly one monitor running per database
+ * it takes an exclusive lock on the CITUS_BACKGROUND_TASK_MONITOR
+ * operation. This lock is consulted from the maintenance daemon to only
+ * spawn a new monitor when the lock is not held.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "safe_mem_lib.h"
+
+#include "access/xact.h"
+#include "commands/dbcommands.h"
+#include "libpq-fe.h"
+#include "libpq/pqformat.h"
+#include "libpq/pqmq.h"
+#include "libpq/pqsignal.h"
+#include "parser/analyze.h"
+#include "pgstat.h"
+#include "storage/dsm.h"
+#include "storage/ipc.h"
+#include "storage/shm_mq.h"
+#include "storage/shm_toc.h"
+#include "tcop/pquery.h"
+#include "tcop/tcopprot.h"
+#include "tcop/utility.h"
+#include "utils/memutils.h"
+#include "utils/portal.h"
+#include "utils/ps_status.h"
+#include "utils/resowner.h"
+#include "utils/snapmgr.h"
+#include "utils/timeout.h"
+
+#include "distributed/background_jobs.h"
+#include "distributed/citus_safe_lib.h"
+#include "distributed/listutils.h"
+#include "distributed/maintenanced.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/metadata_utility.h"
+#include "distributed/shard_cleaner.h"
+#include "distributed/resource_lock.h"
+
+bool BackgroundTaskMonitorDebugDelay = false;
+
+/* Table-of-contents constants for our dynamic shared memory segment. */
+#define CITUS_BACKGROUND_TASK_MAGIC 0x51028081
+#define CITUS_BACKGROUND_TASK_KEY_DATABASE 0
+#define CITUS_BACKGROUND_TASK_KEY_USERNAME 1
+#define CITUS_BACKGROUND_TASK_KEY_COMMAND 2
+#define CITUS_BACKGROUND_TASK_KEY_QUEUE 3
+#define CITUS_BACKGROUND_TASK_NKEYS 4
+
+static BackgroundWorkerHandle * StartCitusBackgroundJobExecuter(char *database,
+																char *user, char *command,
+																dsm_segment **pSegment);
+static void ExecuteSqlString(const char *sql);
+static void ConsumeTaskWorkerOutput(shm_mq_handle *responseq, BackgroundTask *task,
+									bool *hadError, bool *cancelled);
+static void UpdateDependingTasks(BackgroundTask *task);
+
+PG_FUNCTION_INFO_V1(citus_jobs_cancel);
+PG_FUNCTION_INFO_V1(citus_jobs_wait);
+
+
+/* pg_catalog.citus_jobs_cancel(jobid bigint) */
+Datum
+citus_jobs_cancel(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	int64 jobid = PG_GETARG_INT64(0);
+
+	/* Cancel all tasks that were scheduled before */
+	List *pids = CancelTasksForJob(jobid);
+
+	/* send cancellation to any running backends */
+	int pid = 0;
+	const int sig = SIGINT;
+	foreach_int(pid, pids)
+	{
+#ifdef HAVE_SETSID
+		if (kill(-pid, sig))
```

Comment:
permission checks are done when iterating the tasks, don't think we can skip the permission checks there, as to prevent any user to be able to cancel jobs of any other user.

as for `pg_signal_backend`, this function seems to _not_ be exported, atleast as of 15beta3. Do you suggest copying the whole function?

also: we had the discussion yesterday on signalling any process after a failure. seems that `pg_signal_backend` has some guarding against signalling non-postgres processes as well.

Reply:
did not realize it was private, but I guess we could use pg_cancel_backend

Reply:
> also: we had the discussion yesterday on signalling any process after a failure. seems that pg_signal_backend has some guarding against signalling non-postgres processes as well.

Your usage of `BackendPidGetProc` should already take care of this.

Reply:
> Like Nils said during lunch recently, sadly we cannot. Because we also define the pg_cancel_backend symbol in our own C files and due to shared library definition winning during resolve, we cannot call the original postgres version anymore.

we should fix that now-ish

Reply:
Actually this turned out to be fixable: https://github.com/citusdata/citus/pull/6300

Reply:
after merging main I now use `pg_cancel_backend` and `pg_terminate_backend` for these signals.